### PR TITLE
feat: add alchemy transport update for the mav2 account client

### PR DESCRIPTION
**Pull request by Blu-J**
# Pull Request Checklist

- [x] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [x] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [x] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [x] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [x] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

## PR-Codex overview
This PR introduces a new type `NotType` for type safety in TypeScript and modifies various client functions to ensure that specific transport types are not used with certain clients. It also adds tests for the `createSMAV2AccountClient` function, ensuring proper behavior with Alchemy transport.

### Detailed summary
- Added `NotType` type to enforce type constraints.
- Updated parameters in `createLightAccountClient`, `createMultiOwnerLightAccountClient`, `createMultiOwnerModularAccountClient`, and `createMultisigModularAccountClient` to use `NotType`.
- Added tests for `createSMAV2AccountClient` to validate Alchemy transport behavior.
- Enhanced `createSMAV2AccountClient` to differentiate between Alchemy and custom transports.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by Blu-J** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/1271?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1271** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/1271?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/1271?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1264** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/1264?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1248** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/1248?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1236** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/1236?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/alchemyplatform?ref=merge-queue-instructions-in-stack-comment&prId=1271).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: account-kit/smart-contracts/src/ma-v2/client/client.ts [hunk 1]
```diff
@@ -11,10 +11,24 @@ import {
...
-  TSigner extends SmartAccountSigner = SmartAccountSigner
-> = SmartAccountClient<Transport, Chain, MAV2Account<TSigner>>;
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+  TChain extends Chain = Chain
+> = SmartAccountClient<Transport, TChain, MAV2Account<TSigner>>;
+
+export type SMAV2AlchemyAccountClient<
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+  TChain extends Chain = Chain
+> = SmartAccountClient<AlchemyTransport, TChain, MAV2Account<TSigner>>;
```
Comment (moldy530) [inline account-kit/smart-contracts/src/ma-v2/client/client.ts:31]: nit: wonder if we can merge this somehow and infer the type based on a `TTransport extends AlchemyTransport | Transport`

if `TTransport extends AlchemyTransport` -> `SmartAccountClient<AlchemyTransport....`

and same for the params below, kinda how you're doing it [here](https://github.com/alchemyplatform/aa-sdk/pull/1264/files#diff-0c341232c39eadc2fd26bd4ac9bf0a4dfac0731b7d9470bac6a9a8b9d3e82aaaR78)

Comment (moldy530) [reply account-kit/smart-contracts/src/ma-v2/client/client.ts:31]: though you also mentioned at some point that avoiding doing this helped improve readability of the type on hover, is this a good example of that?

File: account-kit/smart-contracts/src/ma-v2/client/client.ts [hunk 2]
```diff
@@ -64,7 +97,21 @@ export function createSMAV2AccountClient<
...
   config: CreateSMAV2AccountClientParams
-): Promise<SmartAccountClient> {
+): Promise<SmartAccountClient | AlchemySmartAccountClient> {
+  const { transport, chain } = config;
+  if (isAlchemyTransport(transport, chain)) {
+    const smaV2Account = await createSMAV2Account({
+      ...config,
+      transport,
+      chain,
+    });
```
Comment (moldy530) [inline account-kit/smart-contracts/src/ma-v2/client/client.ts:107]: nit: we should be able to just pull this up and remove line 115 below since the account instance can be created the same way regardless of the client

File: account-kit/smart-contracts/src/ma-v2/client/client.ts [hunk 3]
```diff
@@ -64,7 +97,21 @@ export function createSMAV2AccountClient<
  */
 export async function createSMAV2AccountClient(
   config: CreateSMAV2AccountClientParams
-): Promise<SmartAccountClient> {
+): Promise<SmartAccountClient | AlchemySmartAccountClient> {
```
Comment (moldy530) [inline account-kit/smart-contracts/src/ma-v2/client/client.ts:100]: for my own understanding: does it complain if this is just `SmartAccountClient`? technically the AlchemySmartAccountClient is a `SmartAccountClient`

**Review by moldy530 — COMMENTED** [review]
couple nits and just some questions for my own understanding but overall LGTM

We should probs also test the types at a minimum that with AlchemyTransport we get an Alchemy client and any other transport we get a non Alchemy Client

File: account-kit/smart-contracts/src/ma-v2/client/client.ts (hunk 1 above)
Comment (Blu-J) [reply account-kit/smart-contracts/src/ma-v2/client/client.ts:31]: Okay, the SMAV2AccountClient is trivial, since the output for the createSMAV2AccountClient is extended and each of the return types can be based on which params passed in on the type overloading. 

Thoughts on the params, it would probably make it ugly and I don't like this
```diff
diff --git a/account-kit/smart-contracts/src/ma-v2/client/client.ts b/account-kit/smart-contracts/src/ma-v2/client/client.ts
index 806104d7..e2cb0585 100644
--- a/account-kit/smart-contracts/src/ma-v2/client/client.ts
+++ b/account-kit/smart-contracts/src/ma-v2/client/client.ts
@@ -26,33 +26,33 @@ export type SMAV2AccountClient<
   TTransport extends Transport | AlchemyTransport = Transport
 > = SmartAccountClient<TTransport, TChain, MAV2Account<TSigner>>;
 
+// prettier-ignore
 export type CreateSMAV2AccountClientParams<
-  TTransport extends Transport = Transport,
+  TTransport extends Transport | AlchemyTransport = Transport,
   TChain extends Chain = Chain,
   TSigner extends SmartAccountSigner = SmartAccountSigner
-> = CreateSMAV2AccountParams<TTransport, TSigner> &
+> = 
+  Omit<
+    CreateSMAV2AccountParams<TTransport, TSigner> &
       Omit<
        SmartAccountClientConfig<TTransport, TChain>,
        "transport" | "account" | "chain"
-  >;
-export type CreateSMAV2AlchemyAccountClientParams<
-  TTransport extends AlchemyTransport = AlchemyTransport,
-  TChain extends Chain = Chain,
-  TSigner extends SmartAccountSigner = SmartAccountSigner
-> = Omit<
-  CreateSMAV2AccountClientParams<TTransport, TChain, TSigner>,
-  "transport"
-> &
-  Omit<
+      >,
+    TTransport extends AlchemyTransport ? "transport" :
+    never
+  > & (
+    TTransport extends AlchemyTransport ? Omit<
         AlchemySmartAccountClientConfig<TChain, LightAccount<TSigner>>,
           "account"
-  >;
+      > :
+    {}
+  );
 
 export function createSMAV2AccountClient<
   TChain extends Chain = Chain,
   TSigner extends SmartAccountSigner = SmartAccountSigner
 >(
-  args: CreateSMAV2AlchemyAccountClientParams<AlchemyTransport, TChain, TSigner>
+  args: CreateSMAV2AccountClientParams<AlchemyTransport, TChain, TSigner>
 ): Promise<SMAV2AccountClient<TSigner, TChain, AlchemyTransport>>;
 
 export function createSMAV2AccountClient<


```

Comment (Blu-J) [reply account-kit/smart-contracts/src/ma-v2/client/client.ts:31]: image

Comment (moldy530) [reply account-kit/smart-contracts/src/ma-v2/client/client.ts:31]: oh oof yea, ok yea let's not merge

File: account-kit/smart-contracts/src/ma-v2/client/client.ts [hunk 4]
```diff
@@ -25,13 +35,34 @@ export type CreateSMAV2AccountClientParams<
...
+  args: CreateSMAV2AlchemyAccountClientParams<AlchemyTransport, TChain, TSigner>
+): Promise<SMAV2AccountClient<TSigner, TChain, AlchemyTransport>>;
+
+export function createSMAV2AccountClient<
+  TTransport extends Transport = Transport,
+  TChain extends Chain = Chain,
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+>(
+  args: CreateSMAV2AccountClientParams<TTransport, TChain, TSigner> &
+    NotType<TTransport, AlchemyTransport>
```
Comment (Blu-J) [inline account-kit/smart-contracts/src/ma-v2/client/client.ts:64]: 📓 So, this here was needed for the test to make sure that we didn't filter on the type alchemyTransport for the other case

File: account-kit/smart-contracts/src/ma-v2/client/client.ts [hunk 5]
```diff
@@ -25,13 +35,34 @@ export type CreateSMAV2AccountClientParams<
...
+  TChain extends Chain = Chain,
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+> = Omit<
+  CreateSMAV2AccountClientParams<TTransport, TChain, TSigner>,
+  "transport"
+> &
+  Omit<
+    AlchemySmartAccountClientConfig<TChain, LightAccount<TSigner>>,
+    "account"
+  > & { paymasterAndData?: never };
```
Comment (Blu-J) [inline account-kit/smart-contracts/src/ma-v2/client/client.ts:49]: 📓 That without this this test fails
image

File: account-kit/smart-contracts/src/ma-v2/client/client.ts [hunk 6]
```diff
@@ -11,10 +12,19 @@ import {
   type CreateSMAV2AccountParams,
   type MAV2Account,
 } from "../account/semiModularAccountV2.js";
-
+import {
+  createAlchemySmartAccountClient,
+  isAlchemyTransport,
+  type AlchemySmartAccountClient,
```
Comment (moldy530) [inline account-kit/smart-contracts/src/ma-v2/client/client.ts:18]: ⚠️ **[eslint]** <[@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars)> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'AlchemySmartAccountClient' is defined but never used.

File: account-kit/smart-contracts/src/ma-v2/client/client.ts (hunk 5 above)
Comment (moldy530) [reply account-kit/smart-contracts/src/ma-v2/client/client.ts:48]: oh nice, I think we also need to add `dummyPaymasterAndData?: never` here as well

**Review by moldy530 — APPROVED** [review]
had one comment, but LGTM otherwise!

**Review by moldy530 — APPROVED** [review]
